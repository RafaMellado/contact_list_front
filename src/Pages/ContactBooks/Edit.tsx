import { useEffect, useState } from "react";
import { useTranslation } from "react-i18next";
import { useNavigate, useParams } from "react-router-dom";
import ContactBooksService from "../../Services/ContactBooksService";
import {
  ContactBook,
  ContactBookRequestBody,
  ContactBookRequestError,
} from "../../Services/Interfaces/ContactBook";
import { ContactBookForm } from "./Components/Form";

export function ContactBookEdit() {
  const [contactBook, setContactBook] = useState<ContactBook>();
  const [errors, setErrors] = useState<ContactBookRequestError>();

  const { id } = useParams();
  const { t } = useTranslation();
  const navigate = useNavigate();

  const TRANSLATIONS: string = "contactBooksEdit";

  const backToHome = () => {
    navigate("/home");
  };

  const editContactBook = async (data: ContactBookRequestBody) => {
    try {
      await ContactBooksService.update(Number(id), data);

      backToHome();
    } catch (errors) {
      setErrors(errors as ContactBookRequestError);
    }
  };

  const getContactBook = async () => {
    const response = await ContactBooksService.show(Number(id));

    setContactBook(response);
  };

  useEffect(() => {
    getContactBook();
  }, [id]);

  return (
    <div className="d-flex flex-column align-items-center">
      {contactBook?.name && (
        <h2 data-testid="contact-book-edit-title" className="mb-4">
          {t<string>(`${TRANSLATIONS}.title`, { name: contactBook.name })}
        </h2>
      )}
      <ContactBookForm
        translations={TRANSLATIONS}
        item={contactBook}
        submitFn={editContactBook}
        backFn={backToHome}
        errors={errors}
      />
    </div>
  );
}
