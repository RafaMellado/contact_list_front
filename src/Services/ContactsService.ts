import {
  Contact,
  ContactRequestBody,
  ContactWithContactHistories,
} from "./Interfaces/Contact";
import ContactsRepository from "./Repository/ContactsRepository";

class ContactsService {
  index(params: object): Promise<Contact[]> {
    return ContactsRepository.index(params);
  }

  show(id: number): Promise<ContactWithContactHistories> {
    return ContactsRepository.show(id);
  }

  create(data: ContactRequestBody): Promise<Contact> {
    return ContactsRepository.create(data);
  }

  update(id: number, data: ContactRequestBody): Promise<Contact> {
    return ContactsRepository.update(id, data);
  }

  delete(id: number): Promise<Response> {
    return ContactsRepository.delete(id);
  }
}

export default new ContactsService();
