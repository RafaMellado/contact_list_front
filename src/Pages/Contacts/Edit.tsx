import { useEffect, useState } from "react";
import { useTranslation } from "react-i18next";
import { useNavigate, useParams } from "react-router-dom";
import ContactsService from "../../Services/ContactsService";
import {
  Contact,
  ContactRequestBody,
  ContactRequestError,
} from "../../Services/Interfaces/Contact";
import { ContactForm } from "./Components/Form";

export function ContactEdit() {
  const [contact, setContact] = useState<Contact>();

  const { id } = useParams();
  const { t } = useTranslation();
  const navigate = useNavigate();
  const [errors, setErrors] = useState<ContactRequestError>();

  const TRANSLATIONS: string = "contactsEdit";

  const backToList = () => {
    navigate(`/contact-book/show/${contact?.contact_book_id}`);
  };

  const editContact = async (data: Partial<ContactRequestBody>) => {
    const newData = {
      ...data,
      contact_book_id: Number(contact?.contact_book_id),
    };

    try {
      await ContactsService.update(Number(id), newData as ContactRequestBody);

      backToList();
    } catch (errors) {
      setErrors(errors as ContactRequestError);
    }
  };

  useEffect(() => {
    ContactsService.show(Number(id)).then((response) => setContact(response));
  }, [id]);

  return (
    <div className="d-flex flex-column align-items-center">
      <h2 className="mb-4">
        {t<string>(`${TRANSLATIONS}.title`, {
          name: `${contact?.givenname} ${contact?.surname}`,
        })}
      </h2>

      <ContactForm
        data={contact}
        translations={TRANSLATIONS}
        errors={errors}
        submitFn={editContact}
        backToListFn={backToList}
      />
    </div>
  );
}
