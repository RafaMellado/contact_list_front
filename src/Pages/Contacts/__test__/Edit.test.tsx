import { ContactEdit } from "../Edit";
import { act, fireEvent, render, waitFor } from "@testing-library/react";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import ContactsService from "../../../Services/ContactsService";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({ t: (key: string) => key }),
}));

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({ id: 10 }),
}));

jest.mock("../../../Services/ContactsService");

const mockedContactService = ContactsService as jest.Mocked<
  typeof ContactsService
>;

describe(`<${ContactEdit.name} />`, () => {
  const history = createMemoryHistory();

  const factoryComponent = async () => {
    const component = render(
      <Router location="/" navigator={history}>
        <ContactEdit />
      </Router>
    );

    await waitFor(() => {
      expect(component.getByTestId("contact-edit-title")).toBeInTheDocument();
    });

    return component;
  };

  const formItemData = {
    givenname: "test",
    surname: "test surname",
    email: "test@email.com",
    phone: "1234",
    contact_book_id: 1,
  };

  beforeEach(() => {
    mockedContactService.show.mockResolvedValue({
      id: 10,
      contact_histories: [],
      ...formItemData,
    });
  });

  test("renders correctly the component", async () => {
    const component = await factoryComponent();

    act(() => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  test("click on backFn btn redirects correctly", async () => {
    const component = await factoryComponent();

    fireEvent.click(component.getByTestId("contact-form-back-btn"));

    expect(history.location.pathname).toBe("/contact-book/show/1");
  });

  test("submit form calls editContact correctly", async () => {
    const component = await factoryComponent();

    fireEvent.submit(component.getByTestId("contact-form"));

    expect(mockedContactService.update).toHaveBeenCalledWith(10, formItemData);
  });

  test("submit form calls editContact correctly and set errors", async () => {
    mockedContactService.update.mockRejectedValue({
      errors: {
        email: [{ error: "email", value: "test" }],
        phone: [{ error: "phone", value: "test" }],
      },
    });

    const component = await factoryComponent();

    fireEvent.submit(component.getByTestId("contact-form"));

    await waitFor(() => {
      expect(component.queryByTestId("error-box-email")).toBeTruthy();
    });

    expect(component.asFragment()).toMatchSnapshot("errors");
  });
});
