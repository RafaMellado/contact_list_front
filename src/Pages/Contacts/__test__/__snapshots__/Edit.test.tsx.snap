// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactEdit /> renders correctly the component 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-column align-items-center"
  >
    <h2
      class="mb-4"
      data-testid="contact-edit-title"
    >
      contactsEdit.title
    </h2>
    <div
      class="form card"
    >
      <div
        class="card-body"
      >
        <form
          class=""
          data-testid="contact-form"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              contactsEdit.givenname
            </label>
            <input
              class="form-control"
              maxlength="24"
              minlength="2"
              placeholder="contactsEdit.givenname"
              required=""
              type="text"
              value="test"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              contactsEdit.surname
            </label>
            <input
              class="form-control"
              maxlength="24"
              minlength="2"
              placeholder="contactsEdit.surname"
              required=""
              type="text"
              value="test surname"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              contactsEdit.email
            </label>
            <input
              class="form-control"
              placeholder="contactsEdit.email"
              required=""
              type="email"
              value="test@email.com"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              contactsEdit.phone
            </label>
            <input
              class="form-control"
              maxlength="12"
              placeholder="contactsEdit.phone"
              required=""
              type="tel"
              value="1234"
            />
          </div>
          <div
            class="d-flex justify-content-between"
          >
            <button
              class="w-100 btn btn-secondary"
              data-testid="contact-form-back-btn"
              type="button"
            >
              contactsEdit.backBtn
            </button>
            <button
              class="w-100 ms-3 btn btn-primary"
              type="submit"
            >
              contactsEdit.submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ContactEdit /> submit form calls editContact correctly and set errors: errors 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-column align-items-center"
  >
    <h2
      class="mb-4"
      data-testid="contact-edit-title"
    >
      contactsEdit.title
    </h2>
    <div
      class="form card"
    >
      <div
        class="card-body"
      >
        <form
          class=""
          data-testid="contact-form"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              contactsEdit.givenname
            </label>
            <input
              class="form-control"
              maxlength="24"
              minlength="2"
              placeholder="contactsEdit.givenname"
              required=""
              type="text"
              value="test"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              contactsEdit.surname
            </label>
            <input
              class="form-control"
              maxlength="24"
              minlength="2"
              placeholder="contactsEdit.surname"
              required=""
              type="text"
              value="test surname"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              contactsEdit.email
            </label>
            <input
              class="form-control"
              placeholder="contactsEdit.email"
              required=""
              type="email"
              value="test@email.com"
            />
            <div
              class="text-danger mt-2 mb-0"
              data-testid="error-box-email"
            >
              <span>
                errors.email
              </span>
            </div>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              contactsEdit.phone
            </label>
            <input
              class="form-control"
              maxlength="12"
              placeholder="contactsEdit.phone"
              required=""
              type="tel"
              value="1234"
            />
            <div
              class="text-danger mt-2 mb-0"
              data-testid="error-box-phone"
            >
              <span>
                errors.phone
              </span>
            </div>
          </div>
          <div
            class="d-flex justify-content-between"
          >
            <button
              class="w-100 btn btn-secondary"
              data-testid="contact-form-back-btn"
              type="button"
            >
              contactsEdit.backBtn
            </button>
            <button
              class="w-100 ms-3 btn btn-primary"
              type="submit"
            >
              contactsEdit.submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
